File: backend/app/services/job_service.py
"""
Business logic for job applications.
"""
from typing import List, Optional
from uuid import UUID
from sqlalchemy.orm import Session
from sqlalchemy import desc
from app.models.job_application import JobApplication, ApplicationStatus
from app.schemas.job_application import (
    JobApplicationCreate,
    JobApplicationUpdate,
    JobApplicationMove,
)


class JobApplicationService:
    """Service for job application operations."""

    @staticmethod
    def get_all(db: Session, user_id: UUID) -> List[JobApplication]:
        """
        Get all job applications for a specific user.

        Args:
            db: Database session
            user_id: User UUID to filter by

        Returns:
            List of job applications ordered by order_index
        """
        return (
            db.query(JobApplication)
            .filter(JobApplication.user_id == user_id)
            .order_by(JobApplication.status, JobApplication.order_index)
            .all()
        )

    @staticmethod
    def get_by_id(db: Session, application_id: UUID, user_id: UUID) -> Optional[JobApplication]:
        """
        Get job application by ID for a specific user.

        Args:
            db: Database session
            application_id: Application UUID
            user_id: User UUID to filter by

        Returns:
            Job application or None
        """
        return (
            db.query(JobApplication)
            .filter(JobApplication.id == application_id, JobApplication.user_id == user_id)
            .first()
        )

    @staticmethod
    def create(db: Session, application: JobApplicationCreate, user_id: UUID) -> JobApplication:
        """
        Create new job application for a specific user.

        Args:
            db: Database session
            application: Application data
            user_id: User UUID who owns this application

        Returns:
            Created job application
        """
        # Get max order_index for the status and user
        max_order = (
            db.query(JobApplication)
            .filter(
                JobApplication.status == application.status,
                JobApplication.user_id == user_id
            )
            .order_by(desc(JobApplication.order_index))
            .first()
        )

        order_index = (max_order.order_index + 1) if max_order else 0

        db_application = JobApplication(
            **application.model_dump(exclude={"order_index"}),
            user_id=user_id,
            order_index=order_index
        )
        db.add(db_application)
        db.commit()
        db.refresh(db_application)
        return db_application

    @staticmethod
    def update(
        db: Session, application_id: UUID, application: JobApplicationUpdate, user_id: UUID
    ) -> Optional[JobApplication]:
        """
        Update job application for a specific user.

        Args:
            db: Database session
            application_id: Application UUID
            application: Updated data
            user_id: User UUID to verify ownership

        Returns:
            Updated job application or None
        """
        db_application = JobApplicationService.get_by_id(db, application_id, user_id)
        if not db_application:
            return None

        update_data = application.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_application, field, value)

        db.commit()
        db.refresh(db_application)
        return db_application

    @staticmethod
    def delete(db: Session, application_id: UUID, user_id: UUID) -> bool:
        """
        Delete job application for a specific user.

        Args:
            db: Database session
            application_id: Application UUID
            user_id: User UUID to verify ownership

        Returns:
            True if deleted, False if not found
        """
        db_application = JobApplicationService.get_by_id(db, application_id, user_id)
        if not db_application:
            return False

        db.delete(db_application)
        db.commit()
        return True

    @staticmethod
    def move(
        db: Session, application_id: UUID, move_data: JobApplicationMove, user_id: UUID
    ) -> Optional[JobApplication]:
        """
        Move application to new status/position (for drag-drop) for a specific user.

        Args:
            db: Database session
            application_id: Application UUID
            move_data: New status and order
            user_id: User UUID to verify ownership

        Returns:
            Updated job application or None
        """
        db_application = JobApplicationService.get_by_id(db, application_id, user_id)
        if not db_application:
            return None

        old_status = db_application.status
        new_status = move_data.status

        # Update status and order
        db_application.status = new_status
        db_application.order_index = move_data.order_index

        # Update stage fields based on status
        if new_status == ApplicationStatus.INTERVIEWING:
            db_application.interview_stage = move_data.interview_stage
            db_application.rejection_stage = None
        elif new_status == ApplicationStatus.REJECTED:
            db_application.rejection_stage = move_data.rejection_stage
            db_application.interview_stage = None
        else:
            db_application.interview_stage = None
            db_application.rejection_stage = None

        # If status changed, reorder other applications (only for this user)
        if old_status != new_status:
            # Increment order_index for applications after the insertion point
            db.query(JobApplication).filter(
                JobApplication.user_id == user_id,
                JobApplication.status == new_status,
                JobApplication.order_index >= move_data.order_index,
                JobApplication.id != application_id,
            ).update(
                {JobApplication.order_index: JobApplication.order_index + 1},
                synchronize_session=False,
            )

        db.commit()
        db.refresh(db_application)
        return db_application

    @staticmethod
    def get_by_status(db: Session, status: ApplicationStatus, user_id: UUID) -> List[JobApplication]:
        """
        Get all applications with a specific status for a specific user.

        Args:
            db: Database session
            status: Application status
            user_id: User UUID to filter by

        Returns:
            List of applications
        """
        return (
            db.query(JobApplication)
            .filter(JobApplication.status == status, JobApplication.user_id == user_id)
            .order_by(JobApplication.order_index)
            .all()
        )

File: backend/src/services/job-service.ts
/**
 * Business logic for job applications
 */
import { eq, and, desc, sql } from 'drizzle-orm';
import { db } from '../db/db.js';
import { jobApplications, companies, type JobApplication, type ApplicationStatus, type Company } from '../db/schema.js';
import type {
  JobApplicationCreate,
  JobApplicationUpdate,
  JobApplicationMove,
} from '@jackdo69/job-tracker-shared-types';
import { ApplicationStatus as ApplicationStatusEnum } from '@jackdo69/job-tracker-shared-types';
import { logger } from '../lib/logger.js';

/**
 * Extended job application with company data
 */
export interface JobApplicationWithCompany extends JobApplication {
  company?: Company | null;
}

/**
 * Get all job applications for a specific user
 * Includes company data if company_id is set
 */
export async function getAllApplications(userId: string): Promise<JobApplicationWithCompany[]> {
  const results = await db
    .select({
      application: jobApplications,
      company: companies,
    })
    .from(jobApplications)
    .leftJoin(companies, eq(jobApplications.companyId, companies.id))
    .where(eq(jobApplications.userId, userId))
    .orderBy(jobApplications.status, jobApplications.orderIndex);

  return results.map((result) => ({
    ...result.application,
    company: result.company,
  }));
}

/**
 * Get job application by ID for a specific user
 * Includes company data if company_id is set
 */
export async function getApplicationById(
  applicationId: string,
  userId: string
): Promise<JobApplicationWithCompany | undefined> {
  const result = await db
    .select({
      application: jobApplications,
      company: companies,
    })
    .from(jobApplications)
    .leftJoin(companies, eq(jobApplications.companyId, companies.id))
    .where(and(eq(jobApplications.id, applicationId), eq(jobApplications.userId, userId)))
    .limit(1);

  if (!result[0]) {
    return undefined;
  }

  return {
    ...result[0].application,
    company: result[0].company,
  };
}

/**
 * Create new job application for a specific user
 */
export async function createApplication(
  applicationData: JobApplicationCreate,
  userId: string
): Promise<JobApplication> {
  try {
    // Get max order_index for the status and user
    const maxOrderResult = await db
      .select({ maxOrder: jobApplications.orderIndex })
      .from(jobApplications)
      .where(and(eq(jobApplications.status, applicationData.status), eq(jobApplications.userId, userId)))
      .orderBy(desc(jobApplications.orderIndex))
      .limit(1);

    const orderIndex = applicationData.orderIndex ?? (maxOrderResult.length > 0 ? (maxOrderResult[0].maxOrder ?? -1) + 1 : 0);

    // Ensure date is a proper Date object
    const applicationDate = typeof applicationData.applicationDate === 'string'
      ? new Date(applicationData.applicationDate)
      : applicationData.applicationDate;

    const insertData = {
      companyId: applicationData.companyId ?? null,
      companyName: applicationData.companyName,
      positionTitle: applicationData.positionTitle,
      status: applicationData.status as ApplicationStatus,
      interviewStage: applicationData.interviewStage ?? null,
      rejectionStage: applicationData.rejectionStage ?? null,
      applicationDate,
      salaryRange: applicationData.salaryRange ?? null,
      location: applicationData.location ?? null,
      notes: applicationData.notes ?? null,
      userId,
      orderIndex,
    };

    logger.debug({ insertData }, 'Attempting to insert job application');

    const [newApplication] = await db
      .insert(jobApplications)
      .values(insertData)
      .returning();

    return newApplication;
  } catch (error) {
    logger.error({
      error: error instanceof Error ? {
        message: error.message,
        stack: error.stack,
        name: error.name
      } : String(error),
      userId,
      applicationData
    }, 'Database error in createApplication');
    throw error;
  }
}

/**
 * Update job application for a specific user
 */
export async function updateApplication(
  applicationId: string,
  applicationData: JobApplicationUpdate,
  userId: string
): Promise<JobApplication | undefined> {
  const existing = await getApplicationById(applicationId, userId);
  if (!existing) {
    return undefined;
  }

  // Build update data (already in camelCase from shared types)
  const updateData: Partial<JobApplication> = {};
  if (applicationData.companyId !== undefined) updateData.companyId = applicationData.companyId;
  if (applicationData.companyName !== undefined) updateData.companyName = applicationData.companyName;
  if (applicationData.positionTitle !== undefined) updateData.positionTitle = applicationData.positionTitle;
  if (applicationData.status !== undefined) updateData.status = applicationData.status as ApplicationStatus;
  if (applicationData.interviewStage !== undefined) updateData.interviewStage = applicationData.interviewStage;
  if (applicationData.rejectionStage !== undefined) updateData.rejectionStage = applicationData.rejectionStage;
  if (applicationData.applicationDate !== undefined) {
    updateData.applicationDate = typeof applicationData.applicationDate === 'string'
      ? new Date(applicationData.applicationDate)
      : applicationData.applicationDate;
  }
  if (applicationData.salaryRange !== undefined) updateData.salaryRange = applicationData.salaryRange;
  if (applicationData.location !== undefined) updateData.location = applicationData.location;
  if (applicationData.notes !== undefined) updateData.notes = applicationData.notes;
  if (applicationData.orderIndex !== undefined) updateData.orderIndex = applicationData.orderIndex;

  const [updated] = await db
    .update(jobApplications)
    .set(updateData)
    .where(and(eq(jobApplications.id, applicationId), eq(jobApplications.userId, userId)))
    .returning();

  return updated;
}

/**
 * Delete job application for a specific user
 */
export async function deleteApplication(
  applicationId: string,
  userId: string
): Promise<boolean> {
  const existing = await getApplicationById(applicationId, userId);
  if (!existing) {
    return false;
  }

  await db
    .delete(jobApplications)
    .where(and(eq(jobApplications.id, applicationId), eq(jobApplications.userId, userId)));

  return true;
}

/**
 * Move application to new status/position (for drag-drop) for a specific user
 */
export async function moveApplication(
  applicationId: string,
  moveData: JobApplicationMove,
  userId: string
): Promise<JobApplication | undefined> {
  const application = await getApplicationById(applicationId, userId);
  if (!application) {
    return undefined;
  }

  const oldStatus = application.status;
  const newStatus = moveData.status as ApplicationStatus;

  // Prepare update data (already in camelCase from shared types)
  const updateData: Partial<JobApplication> = {
    status: newStatus,
    orderIndex: moveData.orderIndex,
  };

  // Update stage fields based on status
  if (newStatus === ApplicationStatusEnum.INTERVIEWING) {
    updateData.interviewStage = moveData.interviewStage ?? null;
    updateData.rejectionStage = null;
  } else if (newStatus === ApplicationStatusEnum.REJECTED) {
    updateData.rejectionStage = moveData.rejectionStage ?? null;
    updateData.interviewStage = null;
  } else {
    updateData.interviewStage = null;
    updateData.rejectionStage = null;
  }

  // If status changed, reorder other applications (only for this user)
  if (oldStatus !== newStatus) {
    // Get applications that need to be shifted
    const appsToShift = await db
      .select()
      .from(jobApplications)
      .where(
        and(
          eq(jobApplications.userId, userId),
          eq(jobApplications.status, newStatus),
          sql`${jobApplications.orderIndex} >= ${moveData.orderIndex}`,
          sql`${jobApplications.id} != ${applicationId}`
        )
      );

    // Increment order_index for each application
    for (const app of appsToShift) {
      await db
        .update(jobApplications)
        .set({ orderIndex: app.orderIndex + 1 })
        .where(eq(jobApplications.id, app.id));
    }
  }

  const [updated] = await db
    .update(jobApplications)
    .set(updateData)
    .where(and(eq(jobApplications.id, applicationId), eq(jobApplications.userId, userId)))
    .returning();

  return updated;
}

/**
 * Get all applications with a specific status for a specific user
 */
export async function getApplicationsByStatus(
  status: ApplicationStatus,
  userId: string
): Promise<JobApplication[]> {
  return db
    .select()
    .from(jobApplications)
    .where(and(eq(jobApplications.status, status), eq(jobApplications.userId, userId)))
    .orderBy(jobApplications.orderIndex);
}
